src/julia.jl

```diff
@@ -33,4 +33,77 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
     multireshape(parent, dimss...)
 end
 
+"""
+    pairwise(f::Function, a[, ::Type{Symmetric}])
+
+Apply the function `f` to all pairwise combinations of elements from `a`. PassPro
+`Symmetric` to only calculate `f(x,y)` when `f(x,y) == f(y,x)` and return a
+`Symmetric Matrix`.
+"""
+function pairwise(f::Function, a)
+    n = length(a)
+    firstval = f(first(a), first(a))
+    r = Matrix{typeof(firstval)}(n, n)
+    pairwise!(f, a, r)
+    r
+end
+
+
+function pairwise(f::Function, a, ::Type{Symmetric})
+    n = length(a)
+    firstval = f(first(a), first(a))
+    r = Symmetric(Matrix{typeof(firstval)}(n, n))
+    pairwise!(f, a, r)
+    r
+end
+
+"""
+    pairwise!(f::Function, a, r)
+
+Apply the function `f` to all pairwise combinations of elements from `a` and
+storing the results in `r`. Setting keyword `symmetric = true` will only calculate
+each variable combination once for symmetric functions. This is useful if `f` is
+expensive.
+"""
+function pairwise!(f::Function, a, r::AbstractMatrix)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r ($(size(r)), should be $((n, n)))"))
+    pairwise_internal!(f, a, r)
+end
+
+function pairwise_internal!(f, a, r::AbstractMatrix)
+    @inbounds for j in eachindex(a)
+        aj = a[j]
+        for i in eachindex(a)
+            r[i,j] = f(a[i], aj)
+        end
+    end
+end
+
+function pairwise_internal!(f, a, r::Symmetric)
+    r.uplo == 'U' ? pairwise_symmetric_U!(f, a, r) : pairwise_symmetric_L!(f, a, r)
+end
+
+function pairwise_symmetric_L!(f, a, r::Symmetric)
+    @inbounds for j in eachindex(a)
+        aj = a[j]
+        i = j
+        while i <= endof(a)
```

Comment:
Doesn't matter much, but do these need to be different functions? Could `pairwise_symmetric_U` and `pairwise_symmetric_L` just pass values for top and bottom indices to the same function that then computes/saves the values?